codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -464,4 +444,125 @@ public Http2Settings pollSentSettings() {
+        @Override
+        public boolean write(int allowedBytes) {
+            if (allowedBytes == 0 && size() != 0) {
+                // No point writing an empty DATA frame, wait for a bigger allowance.
+                return false;
+            }
+            int maxFrameSize = frameWriter().configuration().frameSizePolicy().maxFrameSize();
+            try {
+                int bytesWritten = 0;
+                do {
+                    int allowedFrameSize = Math.min(maxFrameSize, allowedBytes - bytesWritten);
+                    ByteBuf toWrite;
+                    // Let data consume the frame before padding.
+                    if (data.readableBytes() == 0) {
+                        toWrite = Unpooled.EMPTY_BUFFER;
+                    } else  if (data.readableBytes() > allowedFrameSize) {
+                        toWrite = data.readSlice(Math.min(data.readableBytes(), allowedFrameSize)).retain();
+                    } else {
+                        toWrite = data;
+                    }
+                    int writeablePadding = Math.min(allowedFrameSize - toWrite.readableBytes(), padding);
+                    padding -= writeablePadding;
+                    bytesWritten += toWrite.readableBytes() + writeablePadding;
+                    frameWriter().writeData(ctx, streamId, toWrite, writeablePadding,
+                            size == bytesWritten && endOfStream,
+                            size == bytesWritten ? promise : ctx.voidPromise());
+                } while (size != bytesWritten && allowedBytes > bytesWritten);
+                size -= bytesWritten;
+                return true;
+            } catch (Throwable e) {
+                promise.tryFailure(e);
+                ReferenceCountUtil.safeRelease(data);
+                return false;
+            }
+        }
+    }
+
+    /**
+     * Wrap headers so they can be written subject to flow-control. While headers do not have cost against the
+     * flow-control window their order with respect to other frames must be maintained, hence if a DATA frame is
+     * blocked on flow-control a HEADER frame must wait until this frame has been written.
+     */
+    private final class FlowControlledHeaders extends FlowControlledPayloadBase {
+
+        private final Http2Headers headers;
+
+        private FlowControlledHeaders(ChannelHandlerContext ctx, Http2Headers headers,
+                int padding, boolean endOfStream, int streamId, ChannelPromise promise) {
+            super(ctx, padding, endOfStream, streamId, promise);
+            this.headers = headers;
+        }
+
+        @Override
+        public int size() {
+            return 0;
+        }
+
+        @Override
+        public void error(Throwable cause) {
+            promise.tryFailure(cause);
+        }
+
+        @Override
+        public boolean write(int allowedBytes) {
+            frameWriter().writeHeaders(ctx, streamId, headers, padding, endOfStream, promise);
+            return true;
```

Comment:
I dont think so.  If the writer was going to fail synchronously then it should throw, if it's going to fail asynchronously then the future won't have state yet.